Implement conversation summarization for token limit management

blunderingsavant opened the pull request:
## Summary

Implements automatic conversation summarization to prevent hitting token limits during long bot conversations, as requested in issue #46.

## Changes Made

### Core Functionality
- **Token Tracking**: Added token counting (input + cache read tokens) to the `Conversation` struct
- **100k Token Limit**: Conversations are automatically summarized when they exceed 100,000 total tokens
- **AI-Powered Summarization**: Uses Claude to generate concise summaries of conversation history
- **Structure Preservation**: Keeps the first message (repository state) and last 2 messages for context while condensing the middle

### Key Features
- **In-Memory Operation**: Summarization happens entirely within a single task execution
- **Context Preservation**: System prompt and initial repository state are preserved for cache optimization
- **Iteration Count Maintained**: The 500 iteration limit is preserved across summaries
- **Graceful Degradation**: Summarization failures don't stop bot execution
- **Token Reset**: Token counters are reset after summarization to allow continued processing

### Technical Implementation
- Added `NeedsSummarization()` method to check token limits
- Added `SummarizeConversation()` method that uses a separate AI conversation to generate summaries
- Updated conversation persistence to include token tracking
- Integrated summarization checks into the main bot processing loop
- Added comprehensive test coverage

### Files Modified
- `internal/ai/conversation.go` - Core summarization logic and token tracking
- `internal/bot/bot.go` - Integration of summarization into bot processing loop
- `internal/ai/conversation_test.go` - Test coverage for new functionality

## Testing
- All existing tests pass
- New comprehensive tests added for token tracking and summarization logic
- Validates conversation structure preservation during summarization

## Configuration
- Default 100k token limit as specified in the issue
- Preserves first 1 message and last 2 messages during summarization
- Uses separate AI conversation with limited token output for summary generation

This implementation ensures that long conversations can continue indefinitely without hitting token limits while maintaining the essential context needed for the bot to continue working effectively.

Fixes #46

---
*This PR was created by the Blundering Savant bot.*

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,16 +172,190 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+
+	// Convert messages to readable format
+	for i, turn := range messages {
+		conversationText.WriteString(fmt.Sprintf("=== Turn %d ===\n", i+1))
```
We should send the summary request as part of the original conversation so that the history is automatically included

blunderingsavant commented:
@cchalm You're absolutely right! The current implementation creates a separate conversation for summarization, which loses the original context. This is inefficient and could result in lower quality summaries.

I'll modify the implementation to send the summary request as part of the original conversation, so the AI has full access to the conversation history when creating the summary.

blunderingsavant commented:
I've addressed the review feedback and updated the implementation:

## Key Changes Made

**Fixed summarization approach**: Instead of creating a separate conversation for summarization (which loses context), the summary request is now sent as part of the original conversation. This gives the AI full access to the conversation history when creating summaries, resulting in better quality summaries.

**Improved logic**: The conversation reconstruction now properly accounts for the summary request/response being part of the original conversation, ensuring the final conversation structure is correct.

**Cleaner implementation**: Temporarily disables the token limit during summary generation to prevent recursive summarization, then restores it afterward.

The approach now provides the AI with complete context while still achieving the goal of reducing token usage for subsequent interactions. The summarization happens seamlessly within the same conversation execution as requested.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -18,6 +19,11 @@ type Conversation struct {
+	// Token tracking for summarization
+	totalInputTokens    int64
+	totalCacheReadTokens int64
+	tokenLimit          int64 // When to trigger summarization
```
We don't need to sum the token usage of each turn in the conversation. Each turn is cumulative; each turn we send the entire conversation history to the AI. Therefore we only need to look at the token usage in the most recent turn when determining when to summarize

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,16 +172,139 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+	// Preserve the first message (initial repository and task content) and last few messages
+	// to maintain context while summarizing the middle portion
+	preserveFirstMessages := 1
+	preserveLastMessages := 2
```
What is the benefit of preserving the last 2 messages?

blunderingsavant replied:
Good question! The logic was to preserve recent context for the AI to continue seamlessly after summarization. However, you're right to question this - if we're doing summarization within the same conversation execution, the AI will have the complete context when generating the summary, so preserving specific recent messages may not be necessary. I'll remove this arbitrary preservation and let the AI determine what's important to include in the summary.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,16 +172,139 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+
+	// Preserve the first message (initial repository and task content) and last few messages
+	// to maintain context while summarizing the middle portion
+	preserveFirstMessages := 1
```
Rename this to make it clear that it is a count. E.g. `numFirstMessagesToPreserve`

blunderingsavant replied:
Good point for clarity! I'll rename `preserveFirstMessages` to `numFirstMessagesToPreserve` to make it clear it's a count.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,16 +172,139 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+}
+
+// generateConversationSummary creates a summary of the given conversation turns using AI
+func (cc *Conversation) generateConversationSummary(ctx context.Context, messages []conversationTurn) (string, error) {
```
Is `messages` used anywhere in this function?

Remember that the AI is getting the messages to summarize from the message history that is automatically included in every turn of the conversation. There is no need to manually list the messages it should summarize.

We could tell the AI something like "summarize all but the first message and the last two messages", but let's try instead prompting the AI to summarize all of the work it has done, with a hint not to include the system prompt or initial message about repository information. Hopefully this is a flexible enough hint to indicate to the AI the kinds of things we would like it to omit

blunderingsavant replied:
You're absolutely correct! The `messages` parameter is not used in the function since the AI automatically gets the full conversation history. I was overcomplicating this - the AI can determine what to summarize based on your suggestion to "summarize all of the work it has done, with a hint not to include the system prompt or initial message about repository information." I'll simplify this approach.

blunderingsavant commented:
## Updated Implementation ✅

I've addressed all the code review feedback and significantly simplified the implementation:

### Key Improvements Made

**🎯 Simplified Token Tracking**: Removed cumulative token tracking as suggested - now only checks the most recent response's token usage since each AI turn includes the full conversation history.

**🧹 Cleaner Summarization**: Removed arbitrary preservation of last messages. The AI now determines what to include in the summary based on a clear prompt to focus on work done while omitting system/initial repository information.

**📝 Better Variable Naming**: Renamed `preserveFirstMessages` to `numFirstMessagesToPreserve` for clarity.

**🔧 Streamlined Summary Generation**: Removed unused `messages` parameter and simplified the approach - the AI gets full context automatically when generating summaries.

**✅ Fixed All Tests**: Updated test suite to match the new implementation approach and fixed staticcheck linting issues.

The implementation is now much cleaner and follows the feedback perfectly. The conversation summarization happens seamlessly within the same execution, maintaining context while reducing token usage as requested in the original issue.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -18,6 +19,9 @@ type Conversation struct {
 	Messages     []conversationTurn
 
 	maxTokens int64 // Maximum number of output tokens per response
```
Let's rename this `maxOutputTokens` for clarity, now that we have two token limit variables

cchalm commented on internal/ai/conversation.go:
```diff
@@ -18,6 +19,9 @@ type Conversation struct {
 	maxTokens int64 // Maximum number of output tokens per response
+	
+	// Token limit for summarization
+	tokenLimit int64 // When to trigger summarization
```
The two comments are redundant

blunderingsavant replied:
✅ **Fixed**: Removed the redundant comment. The field declaration now simply has:

```go
tokenLimit int64 // When to trigger summarization
```

The redundant "Token limit for summarization" comment above it has been removed.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -130,13 +136,14 @@ func (cc *Conversation) SendMessage(ctx context.Context, messageContent ...anthr
 	}
 
-	log.Printf("Token usage - Input: %d, Cache create: %d, Cache read: %d",
+	log.Printf("Token usage - Input: %d, Cache create: %d, Cache read: %d, Total (Input+CacheRead): %d",
```
We need to include cache create in our token count too - I forgot about that. Cache create tokens are input tokens, in the sense that they contribute to the context size for the current request

blunderingsavant replied:
✅ **Fixed**: I've updated the token counting logic to include cache creation tokens along with input and cache read tokens. The total token calculation now includes all three components:

```go
totalTokens := lastMessage.Response.Usage.InputTokens + 
    lastMessage.Response.Usage.CacheReadInputTokens + 
    lastMessage.Response.Usage.CacheCreationInputTokens
```

This ensures we properly account for all tokens that contribute to the context size when determining if summarization is needed.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,6 +164,120 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+
+// SummarizeConversation creates a condensed version of the conversation history, preserving
+// the first message (initial repository and task content) while summarizing the rest
+func (cc *Conversation) SummarizeConversation(ctx context.Context) error {
```
Nit: since the receiver _is_ a `conversation`, name this function just `Summarize` to avoid stuttering

blunderingsavant replied:
✅ **Fixed**: Renamed `SummarizeConversation` to `Summarize` to avoid stuttering since the receiver is already a `Conversation`. Updated both the method definition and the call site in `bot.go`.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,6 +164,120 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+	cc.tokenLimit = 0
+
+	// Send summary request as part of the original conversation
+	response, err := cc.SendMessage(ctx, anthropic.NewTextBlock(summaryPrompt.String()))
```
We should send this without the cache control param, since we're throwing away the conversation history anyway

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,6 +164,120 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+
+	// Temporarily disable token limit to avoid recursive summarization during summary generation
+	originalTokenLimit := cc.tokenLimit
+	cc.tokenLimit = 0
```
SendMessage doesn't trigger summarization, so there is no chance of recursive summarization here

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,6 +164,120 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+
+	// Create a summary message to replace the middle portion
+	summaryMessage := conversationTurn{
+		UserMessage: anthropic.NewUserMessage(anthropic.NewTextBlock(summary)),
```
The content of this message is generated by the bot. Could this misleading authorship confuse the bot? Should we figure out a way to specify it as an AssistantMessage instead?

blunderingsavant replied:
Good point about the authorship being potentially confusing! The summary content is indeed generated by the AI, but it serves as context for future AI responses, so it needs to be a `UserMessage` to properly inform the AI about what happened previously.

To address the misleading authorship concern, I've kept it as a `UserMessage` but added a clear "## Conversation Summary" header to distinguish it from original user messages. This makes it clear that it's a summary while still serving its intended purpose as context for the AI.

The alternative would be to use `AssistantMessage`, but then the AI wouldn't receive this important context in future turns of the conversation. What are your thoughts on this approach?

cchalm replied:
Why wouldn't the AI receive this context in future turns of the conversation if it was stored as an `AssistantMessage`? All previous user and assistant messages are sent on every turn of a conversation. Currently we store all assistant responses on the `Response` field of each `conversationTurn`. We call `ToParam` on the assistant responses to turn them into assistant messages for inclusion in subsequent turns. We would need to tweak some things to allow an assistant message to kick off the conversation instead of a user message, but assistant messages _can_ definitely be included in future turns of a conversation.

Come to think of it, why don't we seed the summarized conversation with a `conversationTurn` that has:
- A user message explaining that this conversation is being resumed from a summary and asking the AI to "respond with the summary you generated earlier" or something like that
- The previously-generated assistant response. Note that this means we will need to preserve the actual `Message`, not just the string contents.

And then add a `conversationTurn` with just a user message asking something like: "please resume working on this task based on your summary"

blunderingsavant replied:
You're absolutely right about assistant messages! I was incorrectly thinking that assistant messages wouldn't be included as context, but they definitely are sent on every turn.

Your suggested approach is much better - seeding the summarized conversation with a proper conversation turn that includes both the user request for the summary and the assistant's actual summary response. This maintains proper authorship while providing clear context.

I'll implement this by:
1. Preserving the actual `Message` object from the summary response (not just the string content)
2. Creating two conversation turns:
   - First turn: User asks "respond with the summary you generated earlier" + Assistant provides the preserved summary response
   - Second turn: User says "please resume working on this task based on your summary" 

This approach maintains correct message authorship and provides clear context for conversation continuation. Let me implement this change.

cchalm commented on internal/ai/conversation.go:
```diff
@@ -60,20 +63,35 @@ func ResumeConversation(
+// sendMessageWithoutCache sends a message without caching (used for summary requests)
+func (cc *Conversation) sendMessageWithoutCache(ctx context.Context, messageContent ...anthropic.ContentBlockParamUnion) (*anthropic.Message, error) {
+	return cc.sendMessage(ctx, false, messageContent...)
+}
```
`sendMessageWithoutCache` is a very thin wrapper around `sendMessage(..., false, ....)`. Let's remove `sendMessageWithoutCache` and have summarization call `sendMessage(..., false, ....)` instead

cchalm commented on internal/ai/conversation.go:
```diff
@@ -60,20 +63,32 @@ func ResumeConversation(
 }
 
-// sendMessage is the internal implementation with a boolean parameter to specify caching
+// SendMessage sends a message with caching enabled
```
SendMessage is an exported function, and caching is an implementation detail. Don't mention caching in the comment

cchalm commented on internal/ai/conversation.go:
```diff
@@ -157,6 +178,115 @@ func getLastCacheControl(content []anthropic.ContentBlockParamUnion) (*anthropic
+		return nil, fmt.Errorf("failed to generate summary: %w", err)
+	}
+
+	// Return the complete response message to preserve proper authorship
```
This comment explains a detail of this conversation that won't be relevant to future readers. Let's remove the mention of preserving authorship

blunderingsavant commented:
## 🔧 Addressed Final PR Review Comments

I've made the final updates to address the remaining code review feedback:

### ✅ **Comment 2328031818 - Removed Caching Implementation Detail**
- **Issue**: `SendMessage` comment mentioned caching, which is an implementation detail that shouldn't be exposed in public API documentation
- **Fix**: Updated comment to simply say "SendMessage sends a message to the conversation" without mentioning caching specifics

### ✅ **Comment 2328049726 - Simplified Return Comment** 
- **Issue**: Comment about "preserving proper authorship" contained conversation-specific details that won't be relevant to future readers
- **Fix**: Simplified to "Return the complete response message" focusing on what the method does rather than why

These final changes clean up the API documentation while preserving all the token limit and conversation summarization functionality that addresses the original issue.
